File: docgen/src/lib/nlpEngine.js
// nlpEngine.js

const OPENAI_API_KEY = process.env.NEXT_PUBLIC_OPENAI_API_KEY;

if (!OPENAI_API_KEY) {
  throw new Error("No OpenAI API key found. Please provide the OpenAI API key in the .env.local file.");
}

// Function to detect language based on file extension or content
function detectLanguage(filePath, content) {
  const extension = filePath.split('.').pop().toLowerCase();
  switch (extension) {
    case 'js':
    case 'jsx':
    case 'ts':
    case 'tsx':
      return 'JavaScript/TypeScript';
    case 'py':
      return 'Python';
    case 'java':
      return 'Java';
    case 'html':
    case 'htm':
      return 'HTML';
    case 'css':
      return 'CSS';
    default:
      // Guess language based on content if extension is not recognized
      if (content.includes('function') || content.includes('const') || content.includes('let')) {
        return 'JavaScript/TypeScript';
      } else if (content.includes('def ') || content.includes('import ')) {
        return 'Python';
      } else if (content.includes('public class') || content.includes('private')) {
        return 'Java';
      } else {
        return 'Unknown';
      }
  }
}

// Function to generate an explanation using the OpenAI API
async function generateExplanation(filePath, content) {
  const language = detectLanguage(filePath, content);
  const apiUrl = "https://api.openai.com/v1/chat/completions";
  
  const body = {
    model: "gpt-3.5-turbo",
    messages: [
      { role: "system", content: `You are a helpful assistant that explains ${language} code.` },
      { role: "user", content: `Explain the following ${language} code in simple terms, focusing on its purpose and main components:\n\n${content}` }
    ],
    max_tokens: 150,
    temperature: 0.7
  };
  
  try {
    const response = await fetch(apiUrl, {
      method: "POST",
      headers: {
        "Authorization": `Bearer ${OPENAI_API_KEY}`,
        "Content-Type": "application/json"
      },
      body: JSON.stringify(body)
    });
    
    if (response.ok) {
      const data = await response.json();
      return data.choices[0].message.content.trim();
    } else {
      throw new Error(`OpenAI API request failed with status: ${response.status}`);
    }
  } catch (error) {
    return `Error generating explanation: ${error.message}`;
  }
}

// Function to generate contextual explanations for multiple files
async function generateContextualExplanations(codeData) {
  const enhancedDocs = [];

  for (const file of codeData) {
    const content = file.content;
    const explanation = await generateExplanation(file.file_path, content);
    
    enhancedDocs.push({
      file_path: file.file_path,
      content: content,
      explanation: explanation
    });
  }
  
  return enhancedDocs;
}

export { generateContextualExplanations, detectLanguage, generateExplanation };

File: nlp_engine.py
from openai import OpenAI
import os
from dotenv import load_dotenv

# Load environment variables from .env file
load_dotenv()

# Initialize the OpenAI client
client = OpenAI(api_key=os.getenv("OPENAI_API_KEY"))

if not client.api_key:
    raise ValueError("No OpenAI API key found. Please set the OPENAI_API_KEY in your .env file.")

def detect_language(file_path, content):
    extension = os.path.splitext(file_path)[1].lower()
    if extension in ['.js', '.jsx', '.ts', '.tsx']:
        return 'JavaScript/TypeScript'
    elif extension in ['.py']:
        return 'Python'
    elif extension in ['.java']:
        return 'Java'
    elif extension in ['.html', '.htm']:
        return 'HTML'
    elif extension in ['.css']:
        return 'CSS'
    else:
        # Try to guess based on content if extension is not recognized
        if 'function' in content or 'const' in content or 'let' in content:
            return 'JavaScript/TypeScript'
        elif 'def ' in content or 'import ' in content:
            return 'Python'
        elif 'public class' in content or 'private' in content:
            return 'Java'
        else:
            return 'Unknown'

def generate_explanation(file_path, content):
    language = detect_language(file_path, content)
    try:
        response = client.chat.completions.create(
            model="gpt-3.5-turbo",
            messages=[
                {"role": "system", "content": f"You are a helpful assistant that explains {language} code."},
                {"role": "user", "content": f"Explain the following {language} code in simple terms, focusing on its purpose and main components:\n\n{content}"}
            ],
            max_tokens=150,
            n=1,
            stop=None,
            temperature=0.7,
        )
        return response.choices[0].message.content.strip()
    except Exception as e:
        return f"Error generating explanation: {str(e)}"

def generate_contextual_explanations(code_data):
    enhanced_docs = []

    for file in code_data:
        content = file['content']
        explanation = generate_explanation(file['file_path'], content)
        
        enhanced_docs.append({
            'file_path': file['file_path'],
            'content': content,
            'explanation': explanation
        })
    
    return enhanced_docs
